runtime/value.go

```diff
@@ -0,0 +1,355 @@
+	v.Lock()
+	if v.status != valueStopped {
+		v.Unlock()
+		return errValueNotStopped
+	}
+	watch, err := v.store.Watch(v.key)
+	if err != nil {
+		v.Unlock()
+		return err
+	}
+	v.watch = watch
+	v.status = valueStarted
+	v.Unlock()
+
+	select {
+	case <-watch.C():
+	case <-time.After(v.opts.InitWatchTimeout()):
+		err = errInitWatchTimeout
+	}
+
+	if err == nil {
+		err = v.updateFn(watch.Get())
+	}
+
+	// NB(xichen): we want to start watching updates even though
+	// we may fail to initialize the value temporarily (e.g., during
+	// a network partition) so the value will be updated when the
+	// error condition is resolved.
+	v.updateWg.Add(1)
+	go v.watchUpdates(watch)
+
+	return err
+}
+
+func (v *value) Update(value kv.Value) error {
+	v.RLock()
+	if v.status == valueClosed {
+		v.RUnlock()
+		return errValueClosed
+	}
+	initialized := v.initialized
+	v.RUnlock()
+
+	if !initialized {
+		return v.initUpdate(value)
+	}
+	return v.cachedUpdate(value)
+}
+
+func (v *value) Stop() error {
+	v.Lock()
+	if v.status != valueStarted {
+		v.Unlock()
+		return errValueNotStarted
+	}
+	if v.watch != nil {
+		v.watch.Close()
+	}
+	v.watch = nil
+	v.status = valueStopping
+	v.Unlock()
+
+	// NB(xichen): we need to release lock while waiting or
+	// otherwise the stopping goroutine and the updating goroutine
+	// may end up in a deadlock situation.
+	v.updateWg.Wait()
+
+	v.Lock()
+	defer v.Unlock()
+
+	// The value may be closed at this point, in which case
+	// we don't override its status.
+	if v.status == valueStopping {
+		v.status = valueStopped
+	}
+	return nil
+}
+
+func (v *value) Close() error {
+	v.Lock()
+	if v.status == valueClosed {
+		v.Unlock()
+		return errValueClosed
+	}
+	if v.watch != nil {
+		v.watch.Close()
+	}
+	v.watch = nil
+	v.status = valueClosed
+	v.Unlock()
+
+	close(v.closedCh)
+	v.updateWg.Wait()
+
+	return nil
+}
+
+func (v *value) watchUpdates(watch kv.ValueWatch) {
+	defer v.updateWg.Done()
+
+	for {
+		select {
+		case _, ok := <-watch.C():
+			if !ok {
+				return
+			}
+			if err := v.updateFn(watch.Get()); err != nil {
+				v.log.Errorf("error updating rules: %v", err)
+			}
+		case <-v.closedCh:
+			return
+		}
+	}
+}
+
+func (v *value) initUpdate(update kv.Value) error {
+	latest, err := v.unmarshalFn(update)
+	if err != nil {
+		return err
+	}
+	newVersion := v.versionFn(latest)
+
+	// If this is the first time we update the rules, we need
+	// to go back in versions until we have found a version
+	// that's already in effect (or have fetched all versions).
+	var (
+		now         = v.nowFn()
+		currVersion = newVersion - 1
+		currCutover = v.cutoverFn(latest)
+		multiErr    = xerrors.NewMultiError()
+		historical  []interface{}
+	)
+
+	for ; currVersion > defaultVersion && currCutover.After(now); currVersion-- {
+		values, err := v.store.History(v.key, currVersion, currVersion+1)
+		if err != nil {
+			err = fmt.Errorf("error getting history for version %d: %v", currVersion, err)
+			multiErr = multiErr.Add(err)
```

Comment:
Is there a way to do indefinite background retries?  I feel like we should be able to support network partitions etc, and the only way to do that is indefinitely in the background be retrying until success.